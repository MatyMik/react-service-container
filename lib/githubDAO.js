let instance;
export default class GithubDAO {
  static getInstance() {
    if (!instance) {
      instance = new GithubDAO();
    }
    return instance;
  }

  async listRepos() {
    const res = await fetch("https://api.github.com/repositories");
    const data = await res.json();
    return data.map((repo) => ({
      id: repo.id,
      name: repo.full_name,
      link: repo.html_url,
    }));
  }
}
