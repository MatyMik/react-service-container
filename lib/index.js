import React, { useContext } from "react";

export const ServiceContainerContext = React.createContext(new Map());

export function ServiceContainer({ providers, children }) {
  const parent = useContext(ServiceContainerContext);
  const container = buildContainer(providers, parent);

  return (
    <ServiceContainerContext.Provider value={container}>
      {children}
    </ServiceContainerContext.Provider>
  );
}

export function useService(serviceToken) {
  const container = useContext(ServiceContainerContext);
  const initFn = container.get(serviceToken);
  return initFn();
}

function buildContainer(providers, parent) {
  const container = new Map(...parent);
  addProviders(container, providers);
  return container;
}

function addProviders(container, providers) {
  normalize(providers).forEach((provider) => {
    if (!("provide" in provider)) {
      throw new Error('Missing "provide" key');
    }

    let initFn;
    switch (true) {
      case "useValue" in provider:
        const value = provider.useValue;
        initFn = () => value;
        break;
      case "useClass" in provider:
        const Ctor = provider.useClass;
        initFn = () => new Ctor();
        break;
      case "useFactory" in provider:
        initFn = provider.useFactory;
        break;
      case "useExisting" in provider:
        if (!container.has(provider.useExisting)) {
          throw new Error("(TODO: Good message about non-existent existing)");
        }
        initFn = container.get(provider.useExisting);
        break;
      default:
        throw new Error("(TODO: Good errors)");
    }

    container.set(provider.provide, initFn);
  });
}

function normalize(providers) {
  return providers.map((provider) => {
    const assumeClassShorthand = !("provide" in provider);
    if (assumeClassShorthand) {
      return { provide: provider, useClass: provider };
    }
    return provider;
  });
}
