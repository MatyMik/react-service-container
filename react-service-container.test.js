import { ServiceContainer, useService } from "./lib";
import React from "react";
import renderer from "react-test-renderer";

class Dep {
  fn() {}
}

const Component = () => {
  const dep = useService(Dep);
  return <p>{dep.fn()}</p>;
};

let mock;
beforeEach(() => {
  mock = jest.fn();
});

test("useValue", () => {
  renderWithProviders([{ provide: Dep, useValue: { fn: mock } }]);
  expect(mock).toHaveBeenCalled();
});

test("useClass", () => {
  class MockClass {
    fn() {
      mock();
    }
  }

  renderWithProviders([{ provide: Dep, useClass: MockClass }]);
  expect(mock).toHaveBeenCalled();
});

test("useFactory", () => {
  const factory = () => ({
    fn: mock,
  });

  renderWithProviders([{ provide: Dep, useFactory: factory }]);
  expect(mock).toHaveBeenCalled();
});

test("useExisting", () => {
  class Existing {
    fn() {
      mock();
    }
  }

  renderWithProviders([
    { provide: Existing, useClass: Existing },
    { provide: Dep, useExisting: Existing },
  ]);
  expect(mock).toHaveBeenCalled();
});

test("Class shorthand", () => {
  class MockDep {
    fn() {
      mock();
    }
  }

  const Component = () => {
    const mockDep = useService(MockDep);
    return <p>{mockDep.fn()}</p>;
  };

  renderer.create(
    <ServiceContainer providers={[MockDep]}>
      <Component />
    </ServiceContainer>
  );
  expect(mock).toHaveBeenCalled();
});

test("Injection Tokens", () => {
  const CONFIG = Symbol.for("config");
  const config = { message: "Hello, world!" };

  const Component = () => {
    const { message } = useService(CONFIG);
    return <p>{message}</p>;
  };
  const component = renderer.create(
    <ServiceContainer providers={[{ provide: CONFIG, useValue: config }]}>
      <Component />
    </ServiceContainer>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test.todo("useExisting with injection tokens");

test.todo("Hierarchal injection");

test.todo("Class components");

test.todo("Descriptive error when missing provider");

test.todo("Descriptive error when provider function throws");

test.todo("Descriptive error message when useExisting lookup fails");

test.todo("Warn when overriding existing provider (in development?)");

function renderWithProviders(providers) {
  return renderer.create(
    <ServiceContainer providers={providers}>
      <Component />
    </ServiceContainer>
  );
}
