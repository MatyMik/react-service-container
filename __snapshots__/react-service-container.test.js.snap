// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Injection Tokens 1`] = `
<p>
  Hello, world!
</p>
`;
